### PR TITLE
Optimize MUTF8String usage

### PR DESCRIPTION
**Pull request by Neakxs**
Decoding was due to debug logging and `''.format()` usage instead of `MUTF8String` methods

```
Runtime : 14.62569200s

              Name               | TotTime  |  Calls   | Source

_create_basic_block              | 2.299055 |    77741 | analysis.py:352   
__init__                         | 1.127192 |   192651 | analysis.py:45    
get_instructions                 | 1.070345 |  1221727 | dvm.py:6411  
__init__                         | 0.979475 |   278862 | dvm.py:4602  
push                             | 0.902524 |  1143986 | analysis.py:176   
__init__                         | 0.715306 |   227703 | dvm.py:5056  
join                             | 0.704111 |   275064 | mutf8.py:113   
get_instruction                  | 0.679435 |  1141095 | dvm.py:6372  
get_instructions                 | 0.539558 |  3040702 | dvm.py:6540  
get_ins_off                      | 0.352696 |     3212 | dvm.py:6610  
<built-in method _struct.unpack> | 0.318943 |  2238376 |
from_bytes                       | 0.258141 |   280156 | mutf8.py:105   
__init__                         | 0.255657 |   159676 | dvm.py:4877  
reload                           | 0.222070 |    82413 | dvm.py:2898  
__init__                         | 0.212520 |    82413 | analysis.py:333   
set_childs                       | 0.200617 |   141011 | analysis.py:157   
get_basic_block                  | 0.190394 |   123176 | analysis.py:239   
__init__                         | 0.175900 |   121167 | dvm.py:4788  
isEnabledFor                     | 0.170479 |   388707 | __init__.py:1515  
__init__                         | 0.161808 |   280156 | mutf8.py:93    
```

I did some changes in `_create_basic_block` in order to reduce loading time.
I still need to check if `Analysis.create_xref()` decode the strings.

Resolves : #694

### CODE REVIEW

**Comment by reox** [comment]
you mean that if you put a slow argument to the log.debug function - even if log.debug does nothing, the slow argument is evaluated, right? Yes, thats obviously a pitfall... We should check if there are more debug statements, which can be annoying. Thanks for finding that!

Are the two PRs ready to merge?

**Comment by Neakxs** [comment]
@reox I think that this PR need more work
Maybe the other one is ready, if you think so

**Comment by reox** [comment]
okay nice!

File: androguard/core/bytecodes/dvm.py [hunk 1]
```diff
@@ -168,9 +168,17 @@ def get_type(atype, size=None):
 
 BREAK_DVM_OPCODES = ["invoke.", "move.", ".put", "if."]
 
-BRANCH_DVM_OPCODES = ["throw", "throw.", "if.", "goto", "goto.", "return",
-                      "return.", "packed-switch$", "sparse-switch$"]
-
+BRANCH_DVM_OPCODES = {
+    "throw" : set([0x27]),
```
Comment (reox) [inline androguard/core/bytecodes/dvm.py]: there is no need to write it like this, I guess.
you can simply write `{0x27}`, which is the same as `set([0x27])`

**Comment by reox** [comment]
@Neakxs can you take a look at the merge in #681 ? I merged in all your recent changes to my changes, and hope I did not break anything...

**Comment by reox** [comment]
I guess bytes is fine!
If it does not work out, we can still change it to str later on.

Is this one now ready to merge?

**Comment by Neakxs** [comment]
You can merge it @reox 

Here are my benchmarks :

Running whole analysis on 6aadac9 :
```
Runtime : 28.34540900s

              Name               | TotTime  |  Calls   | Source

_create_basic_block              | 5.555094 |    77741 | analysis.py:353   
_create_xref                     | 4.627817 |    11335 | analysis.py:1218  
decode                           | 3.518556 |   258574 | mutf8.py:1     
from_bytes                       | 2.346320 |  3733139 | mutf8.py:105   
<method 'match' of '_sre.SRE_Pat | 1.939817 |  9325343 |
bytes                            | 1.850435 | 15847811 | mutf8.py:126   
__init__                         | 1.766998 |  3771326 | mutf8.py:93    
__eq__                           | 1.683603 |  2271604 | mutf8.py:229   
get_name                         | 1.406705 |  9295715 | dvm.py:4042  
__hash__                         | 1.249815 |  4246229 | mutf8.py:201   
__init__                         | 1.182739 |   330950 | analysis.py:46    
__add__                          | 1.041888 |  1724513 | mutf8.py:180   
get_instructions                 | 1.037822 |  1221727 | dvm.py:6382  
join                             | 0.887853 |   700335 | mutf8.py:113   
push                             | 0.876274 |  1143986 | analysis.py:177   
get_basic_block                  | 0.864779 |   241833 | analysis.py:240   
encode                           | 0.854630 |   359401 | mutf8.py:62    
<method 'format' of 'str' object | 0.850622 |   724400 |
get_instructions_idx             | 0.762055 |  1226399 | dvm.py:3074  
get_instructions                 | 0.744497 |  4976587 | dvm.py:6511 
```

Running whole analysis with these changes :
```
Runtime : 18.06118900s

              Name               | TotTime  |  Calls   | Source

_create_xref                     | 4.179459 |    11335 | analysis.py:1222  
_create_basic_block              | 2.881810 |    77741 | analysis.py:359   
<built-in method __new__ of type | 1.764722 |  3971153 |
__hash__                         | 1.572173 |  4246229 | mutf8.py:157   
__init__                         | 1.540342 |   330950 | analysis.py:52    
get_instruction                  | 1.182690 |  1141095 | dvm.py:6344  
__add__                          | 1.135138 |  1678045 | mutf8.py:160   
get_instructions                 | 1.044456 |  4976587 | dvm.py:6512  
get_instructions                 | 0.984186 |  1221727 | dvm.py:6383  
push                             | 0.847957 |  1143986 | analysis.py:183   
get_basic_block                  | 0.835620 |   241833 | analysis.py:246   
__new__                          | 0.804051 |  3971153 | mutf8.py:93    
__init__                         | 0.732277 |  3971153 | mutf8.py:96    
__eq__                           | 0.705375 |  2271604 | mutf8.py:197   
get_instructions_idx             | 0.703638 |  1226399 | dvm.py:3072  
get_byte                         | 0.690019 |  1021791 | dvm.py:203   
join                             | 0.685714 |  1008051 | mutf8.py:111   
get_string                       | 0.568219 |   541882 | dvm.py:7249  
__init__                         | 0.549178 |   278862 | dvm.py:4579  
read                             | 0.541798 |  1775868 | bytecode.py:738 
```

**Comment by reox** [comment]
so next step would be to multithread the xref creation :D 

thanks for the nice optimizations!

**Pull request merged**